crud/src/main/java/com/redhat/lightblue/mediator/Mediator.java

```diff
@@ -594,33 +607,40 @@ public BulkResponse bulkRequest(BulkRequest requests) {
         Error.push("bulk operation");
         ExecutorService executor = Executors.newFixedThreadPool(factory.getBulkParallelExecutions());
         try {
-            LOGGER.debug("Executing {} find requests in parallel", factory.getBulkParallelExecutions());
+            LOGGER.debug("Executing up to {} requests in parallel, ordered = {}", factory.getBulkParallelExecutions(), requests.isOrdered());
             List<Request> requestList = requests.getEntries();
             int n = requestList.size();
             BulkExecutionContext ctx = new BulkExecutionContext(n);
 
             for (int i = 0; i < n; i++) {
                 Request req = requestList.get(i);
-                if (req.getOperation() == CRUDOperation.FIND) {
-                    ctx.futures[i] = executor.submit(getAsyncFind((FindRequest) req));
-                } else {
-                    wait(ctx);
-                    switch (req.getOperation()) {
-                        case INSERT:
-                            ctx.responses[i] = insert((InsertionRequest) req);
-                            break;
-                        case DELETE:
-                            ctx.responses[i] = delete((DeleteRequest) req);
-                            break;
-                        case UPDATE:
-                            ctx.responses[i] = update((UpdateRequest) req);
-                            break;
-                        case SAVE:
-                            ctx.responses[i] = save((SaveRequest) req);
-                            break;
+                if (requests.isOrdered()) {
+                    // ordered - only consecutive finds in parallel
+                    if (req.getOperation() == CRUDOperation.FIND) {
+                        ctx.futures[i] = executor.submit(getFutureRequest(req));
+                    } else {
+                        wait(ctx);
+                        switch (req.getOperation()) {
+                            case INSERT:
+                                ctx.responses[i] = insert((InsertionRequest) req);
+                                break;
+                            case DELETE:
+                                ctx.responses[i] = delete((DeleteRequest) req);
+                                break;
+                            case UPDATE:
+                                ctx.responses[i] = update((UpdateRequest) req);
+                                break;
+                            case SAVE:
+                                ctx.responses[i] = save((SaveRequest) req);
+                                break;
+                        }
                     }
+                } else {
+                    // unordered - do them all in parallel
```

Comment:
For this case, we can merge multiple insert/save requests on the same entity into one. Once the backend returns, we have to process the results to construct the bulk response properly.